### PR TITLE
Enable building DIPY with Meson

### LINKED ISSUE #2514: Reshape our packaging system
Our current system start to be quite old. We might encounter multiple problems in the close future.

Furthermore: 
- `distutils` is deprecated.
- it is not recommended anymore to call directly `setup.py`. See [this article](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html)
- `setup.cfg` will be '_soon_' deprecated in favor of `pyproject.toml` (see the whole setuptools thread https://github.com/pypa/setuptools/issues/1688#issuecomment-851066392)
> the build tools environment currently undergoes both a rapid development and unification. Unification comes mainly from [PEP-517](https://www.python.org/dev/peps/pep-0517/), [PEP-518](https://www.python.org/dev/peps/pep-0518/), [PEP-621](https://www.python.org/dev/peps/pep-0621/) and [PEP-631](https://www.python.org/dev/peps/pep-0631/).

I suppose we should start using `pyproject.toml` and review/clean our current system.

some reading:
--------------

https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html
https://snarky.ca/what-the-heck-is-pyproject-toml/
https://bernat.tech/posts/growing-pain/

### PR DESCRIPTION
**Pull request by skoudoro**
> **`PEP632`: In Python 3.12, distutils will no longer be installed by make install or any of the first-party distribution. Third-party redistributors should no longer include distutils in their bundles or repositories. This PEP makes no specification on migrating the parts of the CPython build process that currently use distutils.**

This is the main motivation of this PR. Python 3.12 should be out in September/October this year. 

Our current build system is using `distutils` 😨  . Since all major packages/depencencies are moving to meson (scipy, numpy, scikit-image, scikit-learn, etc..), we follow the trend 😅 . Futhermore, Meson seems easy to use from my small and short experience. The maintenance should be easier.

**The build should be complete. The full test suite passes.** It is slower for now. I suppose, it is just missing some important compilation flags. Tested only on my macOS M1 (arm64 and rosetta - x86_64).

For now, This contains:
   - A set of meson.build files.
   - pyproject.toml

## To test it

```shell
pip install -r requirements/build.txt
python -m dev.py build
python -m dev.py test -- -svv
```

## Need to be done:

- [ ] add basic documentation
- [x] test on Linux and Windows
- [x] add some CI's
- [x] a more complete pyproject.toml
- [x] Update a lot of compilation flags
- [x] check SIMD support. [this basic module](https://mesonbuild.com/Simd-module.html) should be enough for us
- [x] improve `*.pyd` management. [This will help us a lot](https://github.com/mesonbuild/meson/issues/9049) and it seems to be done. We might need update `cython` and `meson-python` minimum version soon.
- [x] improve `openmp` management
- [x] keep following [meson build system improvements](https://github.com/scipy/scipy/issues/16293) which might help us to simplify the pipeline
- [x] Manage version

## Some References

Quansight Labs blog: [Moving SciPy to the Meson build system](https://labs.quansight.org/blog/2021/07/)
SciPy docs: [How to build SciPy with Meson](https://scipy.github.io/devdocs/dev/contributor/meson.html)
SciPy docs: [advanced Meson build topics](https://docs.scipy.org/doc/scipy/dev/contributor/meson_advanced.html)
[Scipy migration discussion](https://github.com/scipy/scipy/issues/13615)

close #2514

### CODE REVIEW

File: meson.build
Comment (eli-schwartz) [inline meson.build:72]: This NOTE is now in the past, since Meson 0.64.0 was released some time ago and this PR depends on it. SciPy in the meantime has updated this to use `pure: false`, for example. :)

File: meson.build
Comment (eli-schwartz) [inline meson.build:63]: I wonder if you strictly need this? In SciPy's case it's complicated because they have a variety of code sources, and for example, Fortran code doesn't support visibility, so hiding Fortran symbols requires the linker script. And for vendored code embedded as static libraries, it can be too complicated to modify that external code to make it hide itself.

You may be able to get away with not bothering with this. :)

File: MANIFEST.in
Comment (eli-schwartz) [inline MANIFEST.in:13]: Since you're switching the build backend from setuptools to meson-python here, MANIFEST.in should not be needed anymore. It's still usable as `python setup.py install` but it's probably not strictly necessary to teach setuptools how to produce a good meson-based sdist?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

**Comment by skoudoro** [comment]
Thank you for your feedback @eli-schwartz!  I will deeper into all this later this week. As you notice, there is still a lot of cleaning to do. I will ping you back when I feel this PR ready.

**Comment by jarrodmillman** [comment]
We renamed `devpy` to `spin` due to PyPI availability:
- https://github.com/scientific-python/spin
- https://pypi.org/project/spin/

**Comment by stefanv** [comment]
Out of curiosity, what is the status of this PR, and are there any blockers?

**Comment by skoudoro** [comment]
Hi @stefanv,

Thank you for asking.  Currently the main blocker is that the compilation fails on Windows (mingw-64) with this error `fatal error: '-fuse-linker-plugin', but liblto_plugin.dll not found`. Not sure why. Also, we need to test on windows without using mingw-64.

To be correct, they failed with the github action. I did not try yet on a personal window machine (working with macOs)

If you have any idea how to fix this, it will be more than welcomed! 

Otherwise, some optimization flags to check (make sure they are correct in all OS) and make sure that we are using correctly the `*.pxd` file. But those task are not a blocker

**Comment by stefanv** [comment]
Would compiling with clang be an option? That's the only compiler we were able to get working for skimage.

**Comment by skoudoro** [comment]
Interesting, I will get back to this next week and try with clang. 

Thanks @stefanv !

**Comment by skoudoro** [comment]
ok, some progress, it seems to pass on Windows  now with `mingw-w64` from `rtools`. 

Few tests are failing and it might  be due to some compilation flag.

Also, not sure why pytest do not ignore benchmark. the config file is ignored for some reason

tons of commit need to be squash.

File: .spin/cmds.py
Comment (stefanv) [inline .spin/cmds.py:12]: You may also want to look at what we've written for NumPy:

https://github.com/numpy/numpy/blob/main/.spin/cmds.py#L331

Yours is much more straightforward, which I like, but perhaps there are useful features there.

Comment (skoudoro) [reply .spin/cmds.py:12]: that's cool! thanks @stefanv. 

for sure, I will have a closer look later on a new PR. For now, I keep it simple and focus on meson + failing CI's. 

this PR is close to be done. I will tackle / update the documentation on a new PR.

@jarrodmillman and @stefanv, thank you for `spin` tools, really useful, looking forward to add new commands on a new PR. 

ps: Do you know why on Windows `set_num_threads` does not work. it seems that OpenMP use all the threads each times.  Did you encounter this issue?

Comment (stefanv) [reply .spin/cmds.py:12]: Which `set_num_threads` are you referring to?

Are you looking for something like https://pypi.org/project/threadpoolctl/ perhaps?

Comment (skoudoro) [reply .spin/cmds.py:12]: Sorry, I was referring to our own codebase:  https://github.com/dipy/dipy/blob/79c3939e15529edaae6f1da4c8cb92e7abaf4c10/dipy/utils/omp.pyx#L82

but more generally, do you have issue with openmp with scikit-image or other package ?

Comment (stefanv) [reply .spin/cmds.py:12]: We haven't had any reports, but I'm not sure anyone is watching that closely on Windows. I think the most likely place where these issues would have been hashed out is in `threadpoolctl`.

**Comment by skoudoro** [comment]
For now, this PR is done. 

Can you someone test it and have a look @arokem? @Garyfallidis?  this PR superesed #2787.

Next week, I will create new PR:
- to handle our `dipy_*` command-line 
- Update our documentation 
- Update asv management for benchmarking

**Review by jhlegarreta — COMMENTED** [review]
Monumental effort. Thanks @skoudoro and @arokem. An in-line note below.

File: .github/workflows/test.yml
Comment (skoudoro) [reply .github/workflows/test.yml:24]: > I guess 3.12 should be added here? I see that there is a 3.12-pre in the pre workflow, but since 3.8 is dropped, and given that 3.12 is out (and assuming dependencies are already compatible), it may be wise to add it here and transition the pre to 3.13 (if applicable)?

Unfortunately, h5py does not have wheels for 3.12 and compilation failed last time I checked. So, for now, I keep it in pre until they fix it. 

> Wow, dropping 3.8 may impact many users. Not sure if this was anticipated in the previous release, but this is likely to cause unhappiness. I know it may be difficult to plan ahead at times, but IMO communicating this sufficiently in advance would have been necessary (have not found any relevant statement in the release notes).

ok, you go a point here, the end of life is in 1 year, we can keep 6 month more... I forgot the reason of the drop, I need to check again.

Thank you for the feedback !

Comment (jarrodmillman) [reply .github/workflows/test.yml:24]: I believe most of the core packages have already dropped 3.8 and are about to drop 3.9:
https://scientific-python.org/specs/spec-0000/

Comment (skoudoro) [reply .github/workflows/test.yml:24]: Thank you for sharing @jarrodmillman,

So I will keep it as it.

Also, I will read all the documents, and I think we will start following the scientific-python standard and eco-system.  is there anything to do to be part of this scientific-python ecosystem?

Comment (stefanv) [reply .github/workflows/test.yml:24]: DIPY is part of the ecosystem, of course. But following and contributing to development of the SPECs would be great. We are also looking into mechanisms to let projects indicate when they've implemented a SPEC.

Comment (jhlegarreta) [reply .github/workflows/test.yml:24]: Either way, and although a mainenance branch can be opened to backport/aply some critical fixes, IMO users need to be informed beforehand about changrs of this caliber. If we stick to the scientific python conventions, then this needs to be made explicit so that users are aware of it from now on.

Given that this time they were not warned, and unless keeping 3.8 makes things a lot more complicated, I think we should keep it for this release cycle/until its EOL next year.

File: dipy/__init__.py
Comment (arokem) [inline dipy/__init__.py:54]: Would it be possible to automate this with `setuptools.find_packages`?

File: dipy/align/meson.build
Comment (arokem) [inline dipy/align/meson.build:17]: Is this comment still relevant?

**Review by arokem — COMMENTED** [review]
This is great! I had just a couple of really minor comments here.

**Review by arokem — APPROVED** [review]

File: dipy/align/meson.build
Comment (skoudoro) [reply dipy/align/meson.build:17]: Yes, still relevant but not a critic issue. I will dig into it later. 

it is just to detect change on `pxd` file and force a re-compilation. a feature that we did not have before

File: dipy/__init__.py
Comment (skoudoro) [reply dipy/__init__.py:54]: I suppose yes and open to any suggestions

File: pyproject.toml
Comment (jhlegarreta) [inline pyproject.toml:9]: Elefterios : Eleftherios

**Comment by frheault** [comment]
@skoudoro Are the instructions:
```
pip install -r requirements/build.txt
python -m dev.py build
python -m dev.py test -- -svv
```
Still the best way to do it? 

When this is merge, what will the README.md say about local/manual installation?

**Comment by skoudoro** [comment]
No, 

you should be able to do either:

- `pip install dipy` or `pip install .` or
- `pip install -e dipy` or `pip install -e .`   or
- `spin build` or `python -m spin build`
- `spin test` or `python -m spin test`

well, so many option, I need to document. I will recommend to use `spin` because more command will be added like `generate-doc`, `bench`, `release`, .....

**Comment by skoudoro** [comment]
Does anyone know what is the equivalent of [this](https://github.com/dipy/dipy/blob/master/setup.py#L242-L243) in pyproject.toml:

```python
# in setup.py
          data_files=[('share/doc/dipy/examples',
                       glob(pjoin('doc', 'examples', '*.py')))],
```
After a pip install, this specific section goes to `share` folder of the python environment. Maybe it is specific to `setuptools`.

Any idea @jarrodmillman ? @arokem ? @stefanv ? @jhlegarreta ?

**Comment by jhlegarreta** [comment]
Re https://github.com/dipy/dipy/pull/2715#issuecomment-1764647765 I assume you already had a look at this, but just in case:
https://setuptools.pypa.io/en/latest/userguide/datafiles.html

**Comment by eli-schwartz** [comment]
> Does anyone know what is the equivalent of [this](https://github.com/dipy/dipy/blob/master/setup.py#L242-L243) in pyproject.toml:
> 
> ```python
> # in setup.py
>           data_files=[('share/doc/dipy/examples',
>                        glob(pjoin('doc', 'examples', '*.py')))],
> ```
> 
> After a pip install, this specific section goes to `share` folder of the python environment. Maybe it is specific to `setuptools`.

For pyproject.toml, nothing. The pyproject.toml does not define how to install files at all, only build backends do.

As far as installing files to /usr/share goes, in meson you simply install them to `get_option('datadir')`.

**Comment by skoudoro** [comment]
Thank you @eli-schwartz !  it make sense. 

So I will put `python -c "import sysconfig;print(sysconfig.get_config_vars('datarootdir'))"`. in a meson variable and let the user overwrite via `get_option('datadir')`

it should do the job. I will update that later today

**Comment by skoudoro** [comment]
That's perfect, thanks!

ps: I definitively need to return to the documentation of `py_installation` object

File: dipy/meson.build
Comment (eli-schwartz) [inline dipy/meson.build:276]: datadir is already share so you should skip repeating it.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: dipy/meson.build
Comment (skoudoro) [reply dipy/meson.build:276]: This is what I believed too. However, in my conda virtual environment, it does not return the 'share' folder. 

There is something strange... let me try again

Comment (eli-schwartz) [reply dipy/meson.build:276]: You can run `meson configure` in the build directory and it will print out all builtin/project options along with their current values.

Comment (skoudoro) [reply dipy/meson.build:276]: ok, I checked again, still does not work, not sure what is wrong:

- I can see `datadir = share` with `meson configure``
- after using `pip install dipy`, the `share` folder is empty. 
`ls: /opt/homebrew/Caskroom/miniforge/base/envs/dipy-39-x86/share/doc/dipy/examples: No such file or directory`
-  however, I got all my examples in `doc/dipy/examples`.

```
❯ ls /opt/homebrew/Caskroom/miniforge/base/envs/dipy-39-x86/doc/dipy/examples
__pycache__                               reconst_csa.py                            segment_clustering_metrics.py
affine_registration_3d.py                 reconst_csa_parallel.py                   segment_extending_clustering_framework.py
affine_registration_masks.py              reconst_csd.py                            segment_quickbundles.py
afq_tract_profiles.py                     reconst_csd_parallel.py                   simulate_dki.py
brain_extraction_dwi.py                   reconst_cti.py                            simulate_multi_tensor.py
bundle_assignment_maps.py                 reconst_dki.py                            snr_in_cc.py
bundle_extraction.py                      reconst_dki_micro.py                      streamline_formats.py
bundle_group_registration.py              reconst_dsi.py                            streamline_length.py
bundle_registration.py                    reconst_dsi_metrics.py                    streamline_registration.py
bundle_shape_similarity.py                reconst_dsid.py                           streamline_tools.py
bundlewarp_registration.py                reconst_dti.py                            surface_seed.py
cluster_confidence.py                     reconst_forecast.py                       syn_registration_2d.py
combined_workflow_creation.py             reconst_fwdti.py                          syn_registration_3d.py
contextual_enhancement.py                 reconst_gqi.py                            tissue_classification.py
denoise_ascm.py                           reconst_ivim.py                           tracking_bootstrap_peaks.py
denoise_gibbs.py                          reconst_mapmri.py                         tracking_deterministic.py
denoise_localpca.py                       reconst_mcsd.py                           tracking_introduction_eudx.py
denoise_mppca.py                          reconst_msdki.py                          tracking_pft.py
denoise_nlmeans.py                        reconst_qtdmri.py                         tracking_probabilistic.py
denoise_patch2self.py                     reconst_qti.py                            tracking_ptt.py
fast_streamline_search.py                 reconst_qtiplus.py                        tracking_rumba.py
fiber_to_bundle_coherence.py              reconst_rumba.py                          tracking_sfm.py
gradients_spheres.py                      reconst_sfm.py                            tracking_stopping_criterion.py
histology_resdnn.py                       reconst_sh.py                             viz_advanced.py
kfold_xval.py                             reconst_shore.py                          viz_bundles.py
linear_fascicle_evaluation.py             reconst_shore_metrics.py                  viz_roi_contour.py
motion_correction.py                      register_binary_fuzzy.py                  viz_slice.py
path_length_map.py                        reslice_datasets.py                       workflow_creation.py
piesno.py                                 restore_dti.py
quick_start.py                            segment_clustering_features.py
```

Comment (skoudoro) [reply dipy/meson.build:276]: it looks like `get_option('datadir')` return an empty despite `meson configure` giving me the correct default value

Comment (skoudoro) [reply dipy/meson.build:276]:
```
❯ meson --version
1.2.1
```

Comment (eli-schwartz) [reply dipy/meson.build:276]: This appears to be a meson-python bug. It maps "datadir" to /usr instead of /usr/share -- or more generally, to the python prefix instead of python prefix + "/share".

Can you report it to the meson-python bug tracker?

Note that this would work with `spin build`.

Comment (skoudoro) [reply dipy/meson.build:276]: > This appears to be a meson-python bug. It maps "datadir" to /usr instead of /usr/share -- or more generally, to the python prefix instead of python prefix + "/share".

Thank you for your help and finding that

> Can you report it to the meson-python bug tracker?

ok

> Note that this would work with spin build.

I confirm

Comment (skoudoro) [reply dipy/meson.build]: Hi @eli-schwartz, 

What would you recommend as an alternative or "best practices" concerning data files? 

Not sure I want to keep 2 different behaviors between spin and meson-python. I do not think those datafiles should be inside the DIPY package.

Thank you for your feedback.

Comment (eli-schwartz) [reply dipy/meson.build]: I think that it should be relatively simple to fix in meson-python as a long term solution.

One possible "quick hack" is to add a meson_options.txt option called "meson-python" and pass it as a setup arg in the meson-python section of the pyproject.toml -- default it to false and have pip install (via meson-python) set it to true.

Spin will use meson directly and install to `get_option('datadir')`. If the "meson-python" option is truthy then install to `get_option('datadir') / 'share'` with a comment "workaround for meson-python bug: LINK".

Like I said, it's a hack. :D

**Comment by skoudoro** [comment]
Hello everyone! 

any additional comments?  I will wait until Monday evening and if no news, I will go ahead and merge this PR. 

Additional task will be done after this PR merged. I think about:
- Documentation
- Managing CLI
-  updating ASV
- improve version management 
- adding meson_options.txt  if no news from meson-python

**Comment by jhlegarreta** [comment]
> any additional comments? I will wait until Monday evening and if no news, I will go ahead and merge this PR.

Thanks for doing this Serge.

**Comment by skoudoro** [comment]
@jhlegarreta, Can you look why it failed on python3.8? You can see the error here: https://github.com/dipy/dipy/actions/runs/6740916808/job/18324698874?pr=2715#step:7:150

maybe there is another way to get includes files from numpy. thank you.

**Comment by stefanv** [comment]
https://github.com/dipy/dipy/actions/runs/6740916808/job/18324698874?pr=2715#step:9:6644

type float has no attribute sqrt

**Comment by jhlegarreta** [comment]
Python 3.8 failure:
https://github.com/dipy/dipy/actions/runs/6741509003/job/18326222539?pr=2715#step:7:150
```
 ../dipy/meson.build:88:15: ERROR: Command `/Users/runner/work/dipy/dipy/venv/bin/python -c 'import os; os.chdir(".."); import numpy; print(numpy.get_include())'` failed with status 1.
```

At fist sight having to do with mason and maybe the NumPy version in Python 3.8.

**Comment by jhlegarreta** [comment]
@skoudoro concerning https://github.com/dipy/dipy/pull/2715#issuecomment-1792395278 I am not sure if NumPy is available at the time this call is performed (?)
https://github.com/dipy/dipy/actions/runs/6741509003/job/18326222539?pr=2715#step:6:156
https://github.com/dipy/dipy/actions/runs/6741509003/job/18326222539?pr=2715#step:7:9

No trace of messages saying that `numpy` has been installed after the latter line.

**Comment by skoudoro** [comment]
> @skoudoro concerning https://github.com/dipy/dipy/pull/2715#issuecomment-1792395278 I am not sure if NumPy is available at the time this call is performed (?)

Numpy is installed before as you can see here : https://github.com/dipy/dipy/actions/runs/6741509003/job/18326222539?pr=2715#step:6:66

**Comment by jhlegarreta** [comment]
> Numpy is installed before as you can see here : https://github.com/dipy/dipy/actions/runs/6741509003/job/18326222539?pr=2715#step:6:66

OK, missed that line then. Thanks.

**Comment by rgommers** [comment]
Getting the NumPy include directory is a bit cumbersome right now I'm afraid. I suggest to copy exactly what SciPy does: https://github.com/scipy/scipy/blob/6ce4aaf4bc5222e3d1355cd8a392e88f45b4f959/scipy/meson.build#L30-L82.

The different cases are needed to deal with things like in-package-tree virtualenvs. I aim to make this as simple as
```meson
np_dep = dependency('numpy')
```
but that won't land before NumPy 2.0 early next year.

**Comment by skoudoro** [comment]
> Getting the NumPy include directory is a bit cumbersome right now I'm afraid. I suggest to copy exactly what SciPy does:

Thank you @rgommers, looking forward for the new way to handle this!

**Comment by skoudoro** [comment]
Do you have an idea @rgommers why numpy include dir could fail on python3.8. 

see here: https://github.com/dipy/dipy/actions/runs/6786880602/job/18448417652?pr=2715#step:7:159

all our python 3.8 build failed with this same error:

```
  ../dipy/meson.build:106:17: ERROR: Command `/home/runner/work/dipy/dipy/venv/bin/python -c '
  import numpy as np
  try:
    incdir = os.path.relpath(np.get_include())
  except Exception:
    incdir = np.get_include()
  print(incdir)
    '` failed with status 1.
```

**Comment by rgommers** [comment]
> all our python 3.8 build failed with this same error:

You have no entry for numpy with python 3.8 in your build dependencies (`requires = [`) in `pyproject.toml`, so it doesn't get installed for 3.8.

Also, your `pyproject.toml` contains `requires-python = ">=3.9"`. It seems like you should either drop the Python 3.8 CI jobs (which is certainly fine, NumPy and SciPy have already done that) or add in 3.8-specific metadata.

**Comment by skoudoro** [comment]
Arf, how I missed that... Good catch!  thank you @rgommers 

ok, I will just drop the python 3.8 CI's. Sorry @jhlegarreta (but no worries, wheels will be created for python 3.8 for this release a last time).

**Comment by jhlegarreta** [comment]
>  ok, I will just drop the python 3.8 CI's. Sorry @jhlegarreta (but no worries, wheels will be created for python 3.8 for this release a last time).

If you drop CI's do you have a means to know and let people openly know that wheels for Python 3.8 will be working correctly? If the changes Ralf has mentioned are not hard, I would keep 3.8 until the release has been done.

**Comment by skoudoro** [comment]
wheels will not be created if CI's fails for python 3.8. I will let you look at https://github.com/MacPython/dipy-wheels to go deeper in the wheels creation mechanism for DIPY. 

Another thing that we need to document 😅  and clarify.

**Comment by jhlegarreta** [comment]
> wheels will not be created if CI's fails for python 3.8.

Then it seems counter-intuitive to remove the CIs for Python 3.8: if CI's pass, you have a better guarantee that you will not run into trouble creating the wheels. I still do not see why we should drop the CIs it at this stage; it would be 2 lines to modify in `pyproject.toml`.

**Comment by skoudoro** [comment]
ok, I see no additional comment, all is fixed. I am going ahead and merge this PR to move forward for the release. 

Feel free to create an issue if there are any concerns! 

Thanks everyone for the helps / advise / reviews

**Comment by arokem** [comment]
Monumental work on this serge! Kudos and big thanks to everyone who helped!

**Comment by jhlegarreta** [comment]
@skoudoro I have noticed that adapting some of the CI scripts to this change was inadvertently left behind, e.g.
https://github.com/dipy/dipy/blob/master/tools/ci/install.sh#L31

We no longer have a `setup.py` file.

**Pull request merged**
